pkg/registry/generic/rest/proxy.go

```diff
@@ -76,12 +76,33 @@ func (h *UpgradeAwareProxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Re
 		return
 	}
 
+	loc := *h.Location
+	loc.RawQuery = req.URL.RawQuery
+
+	// If original request URL ended in '/', append a '/' at the end of the
+	// of the proxy URL
+	if !strings.HasSuffix(loc.Path, "/") && strings.HasSuffix(req.URL.Path, "/") {
+		loc.Path += "/"
+	}
+
+	// From pkg/apiserver/proxy.go#ServeHTTP:
+	// Redirect requests with an empty path to a location that ends with a '/'
+	// This is essentially a hack for https://github.com/GoogleCloudPlatform/kubernetes/issues/4958.
+	// Note: Keep this code after tryUpgrade to not break that flow.
+	if len(loc.Path) == 0 {
+		var queryPart string
+		if len(req.URL.RawQuery) > 0 {
+			queryPart = "?" + req.URL.RawQuery
+		}
+		w.Header().Set("Location", req.URL.Path+"/"+queryPart)
```

Comment:
Well, this redirection is including the part of the URL that got the user to apiserver's proxy. So it should have everything the user just used to get here.

E.g. the user asked for ```https://apiserver.com/api/v1beta3/namespaces/default/services/foo/proxy?param=blah``` and we want to redirect to ```https://apiserver.com/api/v1beta3/namespaces/default/services/foo/proxy/?param=blah```, not ```apiserver.com/api/v1beta3/namespaces/default/services/foo/proxy/?param=blah```.